### ISSUE TITLE
[Question] Possible to not learn all dimensions of inducing points?

### ISSUE THREAD

**Issue by egibug** [cornellius-gp/gpytorch#2268]
Hi and thank you very much for the great work you've put into this library!

I am trying to train an SVGP where I use an IndexKernel on certain dimensions which are integer encoded. This makes it so that I cannot use `learn_inducing_locations = True` in the `VariationalStrategy` as it seems that this will try to optimize the encoded dimensions as well. These dimensions might then suddenly not be valid for the IndexKernel anymore.

My question is this: is it possible to "mask" certain dimensions of the inducing points from being updated?

I tried to follow this discussion #1598 for a quick and dirty attempt just to see what would happen:

```
class _VariationalStrategy(Module, ABC):
    """
    Abstract base class for all Variational Strategies.
    """

    has_fantasy_strategy = False

    def __init__(self, model, inducing_points, encoded_dim, variational_distribution, learn_inducing_locations=True):
        super().__init__()

        # Model
        object.__setattr__(self, "model", model)

        # Inducing points
        inducing_points = inducing_points.clone()
        if inducing_points.dim() == 1:
            inducing_points = inducing_points.unsqueeze(-1)
        else:
            self.encoded_column = inducing_points[:, encoded_dim]
            columns = [*range(inducing_points.size(1))]
            columns.remove(encoded_dim)
            inducing_points = inducing_points[:, columns]
        if learn_inducing_locations:
            self.register_parameter(name="raw_inducing_points", parameter=torch.nn.Parameter(inducing_points))
        else:
            self.register_buffer("inducing_points", inducing_points)

        # Variational distribution
        self._variational_distribution = variational_distribution
        self.register_buffer("variational_params_initialized", torch.tensor(0))

    @property
    def inducing_points(self):
        return torch.cat((self.raw_inducing_points, self.encoded_column.unsqueeze(-1)), 1)
```

I remove the encoded dimensions from the inducing points being registered as parameters, and then use an `inducing_points` property where I add them back in. But when I try to predict with the model I get an exception since it was expecting less dimensions than in my input.

I'm not very experienced in the inner workings of PyTorch (or GPyTorch) so my approach here might be really naive. I appreciate any help you could give me!

**Comment by egibug** [comment]
It seems like the approach above actually works. I had just forgotten to set `learn_inducing_locations=True`

**Comment by egibug** [comment]
I have a quick suggestion for getting an Approximate GP to learn inducing locations when it has an IndexKernel (see the short lived monologue above):

If we're allowed to set the name of the inducing points parameter at the Variational Strategy base class:
```
class _VariationalStrategy(Module, ABC):
    """
    Abstract base class for all Variational Strategies.
    """

    has_fantasy_strategy = False

    def __init__(self, model, inducing_points, variational_distribution, learn_inducing_locations=True, inducing_points_parameter = "inducing_points"):
        super().__init__()

        # Model
        object.__setattr__(self, "model", model)

        # Inducing points
        inducing_points = inducing_points.clone()
        if inducing_points.dim() == 1:
            inducing_points = inducing_points.unsqueeze(-1)
        if learn_inducing_locations:
            self.register_parameter(name=inducing_points_parameter, parameter=torch.nn.Parameter(inducing_points))
        else:
            self.register_buffer(inducing_points_parameter, inducing_points)

        # Variational distribution
        self._variational_distribution = variational_distribution
        self.register_buffer("variational_params_initialized", torch.tensor(0))
```

then we would have the ability to create a VariationalStrategy as such:
```
class VariationalStrategyWithIndex(gpytorch.variational.VariationalStrategy):
    def __init__(self, model, inducing_points, encoded_dim, variational_distribution, learn_inducing_locations=True):
        self.encoded_column = inducing_points[:, encoded_dim]
        columns = [*range(inducing_points.size(1))]
        columns.remove(encoded_dim)            
        inducing_points = inducing_points[:, columns]

        super().__init__(model, inducing_points, variational_distribution, learn_inducing_locations, inducing_points_parameter = "raw_inducing_points")

    @property
    def inducing_points(self):
        return torch.cat((self.raw_inducing_points, self.encoded_column.unsqueeze(-1)), 1)
```

which allows for learning all dimensions of the inducing points except for those dimensions that are supposed to be integer values.

Not the most intuitive approach, but an easy fix at least. Thoughts?

**Comment by gpleiss** [comment]
@egibug I don't think it's a wise idea to use inducing points when you have discrete-valued dimensions - there are too many headaches involved. The model you are proposing is exactly equivalent to a multitask GP model, where each task corresponds to one value of the discrete feature.

Check out the variational multitask example for more info: https://docs.gpytorch.ai/en/stable/examples/04_Variational_and_Approximate_GPs/SVGP_Multitask_GP_Regression.html
